libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs

```diff
@@ -1229,623 +1230,1733 @@ private static (object, string) ParseISOTimestamp(string timeStamp, Func<DateTim
-                            {
-                                if (count < 0 || count > list.Count)
-                                {
-                                    error = $"{countExpr}={count} which is out of range for {arr}";
-                                }
-                                else
-                                {
-                                    result = arr.ToString().Substring(0, count);
-                                }
-                            }
-                        }
-                    }
+                    destanationTS = TimeZoneInfo.ConvertTimeBySystemTimeZoneId(ts, sourceTimeZone, destinationTimeZone);
+                }
+                else if (ts.Kind == DateTimeKind.Local)
+                {
+                    var utcTS = TimeZoneInfo.ConvertTimeToUtc(ts, TimeZoneInfo.Local);
+                    destanationTS = TimeZoneInfo.ConvertTimeFromUtc(utcTS, destinationTZ);
                 }
                 else
                 {
-                    error = $"{expression.Children[0]} is not array or string.";
+                    destanationTS = TimeZoneInfo.ConvertTimeFromUtc(ts, destinationTZ);
+                }
+
+                try
+                {
+                    result = destanationTS.ToString(format);
+                }
+                catch
+                {
+                    error = "illegal format representation";
                 }
             }
 
             return (result, error);
         }
 
-        private static (object, string) SubArray(Expression expression, object state)
+        private static (object, string) StartOfDay(string timestamp, string format = "o")
         {
-            object result = null;
-            string error;
-            object arr;
-            (arr, error) = expression.Children[0].TryEvaluate(state);
+            string result = null;
+            dynamic ts = null;
+            string error = null;
+            try
+            {
+                ts = DateTime.Parse(timestamp).ToUniversalTime();
+            }
+            catch
+            {
+                error = "illegal timestamp format";
+            }
 
             if (error == null)
             {
-                if (TryParseList(arr, out var list))
+                var startOfDay = ts.Date;
+                try
                 {
-                    var startExpr = expression.Children[1];
-                    object startObj;
-                    (startObj, error) = startExpr.TryEvaluate(state);
-                    var start = 0;
-                    if (error == null)
-                    {
-                        if (startObj == null || !startObj.IsInteger())
-                        {
-                            error = $"{startExpr} is not an integer.";
-                        }
-                        else
-                        {
-                            start = (int)startObj;
+                    result = startOfDay.ToString(format);
```

Comment:
extract ToString is more suitable, reduce hard code error 'illegal format representation'